Linking to External Overlapping Data Sources (WZDz, CWZ, MDS, GBFS, GTFS, Curb Objects, etc)

schnuerle opened the pull request:
## Explain pull request

Adding a reference to WZDx data feeds in CDS's Curbs API per issue #74 

## Is this a breaking change

* No, not breaking

## Impacted Spec

Which API(s) will this pull request impact?

* `Curbs`
* `Events`

schnuerle committed:
Added WZDx reference

jlarsonOmahaNE commented on curbs/README.md:
```diff
@@ -503,7 +508,20 @@ A Location Reference is a JSON object with the following fields:
+| Name   | Type   | Required/Optional   | Description   |
+| ------ | ------ | ------------------- | ------------- |
+| `wzdx_feed_url` | URL | Required | An identifier for the source of the publicly accessible WZDx feed. This MUST be a full HTTPS URL pointing to the data feed which contains more information about the underlying work zone impacting the CDS place. |
+| `wzdx_id` | String | Required | The relevant `id` in the WZDx data feed that impacts the CDS place issued by the data feed provider to identify the WZDx road event, from the [RoadEventFeature](https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/objects/RoadEventFeature.md#roadeventfeature-object-geojson-feature). In case multiple road event `wzdx_id`s are needed, then they would be pro ...
```
@j-d-b I am in favor of using `wzdx_road_event_feature_id` and using "The `id` of a WZDx [RoadEventFeature](https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/objects/RoadEventFeature.md) that impacts the use of a curb zone".

schnuerle replied:
Agree with you all and I updated the field name and description. Take a look please.

j-d-b commented on curbs/README.md:
```diff
@@ -53,6 +53,8 @@ There are four different endpoints that are part of the Curbs API:
     - [Time Span](#time-span)
     - [Rate](#rate) 
   - [Location Reference](#location-reference)
+  - [WXDz Reference](#wzdx-reference)
```
Typo, `WXDz` should be `WZDx`

schnuerle replied:
Thanks for the catch, fixed!

j-d-b commented on curbs/README.md:
```diff
@@ -503,7 +508,20 @@ A Location Reference is a JSON object with the following fields:
+
+## WZDx Reference
+
+A WZDx Reference object defines an external connection to a [Work Zone Data Exchange](https://github.com/usdot-jpo-ode/wzdx) data feed for a given curb place. This allows CDS users to reference public WZDx feeds that impact the area, and see full details in the agency data feed.
```
> defines an external connection to a Work Zone Data Exchange data feed

I think it should be considered that as written, the WZDx Reference doesn't just define the connection to the feed, it describes a specific WZDx road event feature that is relevant to the curb zone.

schnuerle replied:
I made an update to incorporate your suggestions, let me know what you think.

j-d-b replied:
Looks good!

schnuerle committed:
WZDx suggested updates

schnuerle committed:
Update WZDx ID description

natedeshmukhtowery commented:
I think this is in good shape and continues to make sense in aligning CDS and WZDx for relevant work zones that impact curb areas.

@jlarsonOmahaNE I suggest you move forward with a PR to add a corollary optional data element in WZDx to reference CDS ids.

schnuerle commented:
There is now a PR in WZDx to add references to CDS here https://github.com/usdot-jpo-ode/wzdx/pull/326

schnuerle commented on curbs/README.md:
```diff
@@ -503,7 +508,20 @@ A Location Reference is a JSON object with the following fields:
+| Name   | Type   | Required/Optional   | Description   |
+| ------ | ------ | ------------------- | ------------- |
+| `wzdx_feed_url` | URL | Required | An identifier for the source of the publicly accessible WZDx feed. This MUST be a full HTTPS URL pointing to the data feed which contains more information about the underlying work zone impacting the CDS place. |
+| `wzdx_road_event_feature_id` | String | Required | The `id` of a WZDx [RoadEventFeature](https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/objects/RoadEventFeature.md#roadeventfeature-object-geojson-feature) that impacts the use of a curb zone. In case multiple road event `wzdx_id`s are needed, it would be provided in another object with another `wzdx_feed_url`. |
```
Yeah this should be updated to wzdx_road_event_feature_ids and made into an array of Strings. I'll make the edit.

schnuerle replied:
Done

schnuerle commented:
Approved as part of WZDx v4.2 

https://github.com/usdot-jpo-ode/wzdx/blob/release/v4.2/spec-content/objects/CdsCurbZonesReference.md

PR: https://github.com/usdot-jpo-ode/wzdx/pull/326

schnuerle commented:
Does this PR need to be reworked to use Connected Work Zones instead of WZDx terminology and links? @jlarsonOmahaNE

schnuerle committed:
Merge branch 'dev' into feature-wzdx

schnuerle committed:
Made wzdx_road_event_feature_ids an array

mplsmitch commented:
We will be discussing this at the Feb 25th [CDS Working Group meeting](https://github.com/openmobilityfoundation/curb-data-specification/wiki/Web-Conference-2025.02.25-Curb). Please review and leave your comments and questions ahead of time.

natedeshmukhtowery commented:
Would suggest maintaining a reference to both WZDx v4.2 and CWZ. The near-final draft of CWZ maintains the following reference to CDS:

3.6.10 Contents of CdsCurbZonesReference
The CdsCurbZonesReference shall consist of the following requirements. See OpenMobilityFoundation's Curb Data Specification at https://github.com/openmobilityfoundation/curb-data-specification.
a) cds_curb_zone_ids. A list of CDS Curb Zone 'id's. Required.
b) cds_curbs_api_url. An identifier for the source of the requested CDS Curbs API. This MUST be a full HTTPS URL pointing to the main curbs API that contains detailed information about each curb zone identified in 'cds_curb_zone_ids'. Required.

schnuerle commented:
Idea was brought up by steering committee that maybe this can be for any sort of external spec. You would only have to specify the URL and the IDs in that spec where there was a connection to the curb zone.

So an array of 2 items: `feed_url` and `feature_ids`. These could refer to CWZ feed, WZDx 4.1 or 4.2 feed, etc, both or either or any current or future feed.

schnuerle commented:
Wondering if this can be used for MDS overlap as well, eg. MDS trips or states in CDS curb zones/spaces and Events. Going to explore this.

schnuerle committed:
Made data feeds applicable to any standard

schnuerle commented:
We will be talking about this at tomorrow's [Curb Working Group meeting](https://github.com/openmobilityfoundation/curb-data-specification/wiki/Web-Conference-2025.05.20-Curb). We will determine if this solution will work both for CWZ #74 and part of MDS  #17, and an update to support Curb Objects #123.

schnuerle committed:
Create data-types.md

schnuerle committed:
External references link to data objects, includes Curb Objects

schnuerle committed:
Added external references to Events

schnuerle committed:
Added data feed name field

schnuerle committed:
ID description update for data feed

schnuerle commented:
After the WG meeting, the updates requested (to be made by me) are:

- Give examples as a comment in this PR of use in Curbs and in Events so people understand this concept better
- Fix WZDx typo

Would love to have your feedback on this concept @jiffyclub @pierre-bouffort @jacobmalleau @jlarsonOmahaNE @mschwartzie, because it may also be applicable to MDS.

schnuerle committed:
Updated external reference specs and field names and descriptions

schnuerle committed:
Clarify Data Type

schnuerle commented:
May also need the external_references to be in the Curbs Policy object, so that policy can apply to many Curb Zones at once, eg, for a marathon road closure.  

Need to specify that data feeds can be open or authenticated, or possibly have an optional field that states the feed is public or private to make it clear.

jacobmalleau reviewed (commented):
I think this works really well as a generic way to connect feeds, with more relevant linkages being saved for future versions once it is more clear have data feeds are connected. 
I think including a few examples directly in the spec would be helpful. A couple come to mind:
- Note some Curb Zones that also have an MDS policy zone applied to them
- Note a Curb Policy that relates to a WZDX policy.

I also agree this is needed for Curb Policies as well.

schnuerle commented:
Notes from the WG meeting:

- New object to connect a curb zone or a curb event with any external data source
- New data_types.md file, like we have for MDS, with common objects across APIs
- MDS: url, MDS, trip_id, ids - for that Zone. Like Blue Systems in London example parking scooters in MDS trips and CDS Zones.
- Charging: Event happening, can reference CDS Curbs API Object by id. And you can add a OCPI external feed at the same time.
- MDS, CWZ, EV, etc. Any external overlap can be connected
- Seattle - very specific overlap between specs
- CWZ: a driving lane construction area is impacting a CDS Curb zone
- **Add examples** in the PR to help understanding
- SFMTA - would you remove the reference when over? For Events, it's a **moment in time**, recorded. For Curbs, it's **added then removed** when done, if real time is supported. 
- Should reference **be in Curbs Policy object** (not zone?)
- elevate this for more views and feedback, bring it to MDS WG meeting for 2.1 as well.
- Need to specify that data feeds can be open **or** authenticated, or possibly have an **optional field** that states the feed is public or private to make it clear.

schnuerle commented:
More notes on this from [MDS WG meeting](https://github.com/openmobilityfoundation/mobility-data-specification/wiki/Web-conference-notes,-2025.06.05-(MDS-Working-Group)):

- Can align MDS/CDS with any other specs by external reference with a URL and ID connections
- Chicago sees value and will provide feedback
- LADOT sharing screen about Traffic Management Plan TMP - printed and handed out now
- Action item: Make data feed more generic, like a website or document, ids optional

schnuerle committed:
Clarified real time language

schnuerle committed:
Noted when reference applies to Events

schnuerle committed:
Added external reference to Policy

schnuerle committed:
Added field to show if field is public or not

schnuerle committed:
Broadened reference to be any type of URL

schnuerle commented:
Made changes based on requests from both MDS and CDS meetings:

1. Clarified real time language
2. Noted when reference applies to Events
3. Added external reference to Policy
4. Added field to show if field is public or not
5. Broadened reference to be any type of URL
